plugin/ochttp/span_annotator.go

```diff
@@ -0,0 +1,175 @@
+// Copyright 2018, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ochttp
+
+import (
+	"crypto/tls"
+	"net/http"
+	"net/http/httptrace"
+	"strings"
+
+	"go.opencensus.io/trace"
+)
+
+// SpanAnnotator implements the annotation of all available hooks for
+// httptrace.ClientTrace.
+type SpanAnnotator struct {
```

Comment:
I guess it is a matter of opinion on what is the friendlier method for consumers.

Your preference is eliminating the hooks not desired:

```go
func NewSpanAnnotatorWrapper(req *http.Request, s *trace.Span) *httptrace.ClientTrace {
    clientTrace := ochttp.NewSpanAnnotator(req, s)

    // deregister unwanted hooks
    clientTrace.GetConn = nil 
    clientTrace.Got100Continue = nil
    ...
    return clientTrace
}

...

transport := ochttp.Transport{
    NewClientTrace: NewSpanAnnotatorWrapper,
}
```

Mine is choosing the hooks that are:

```go
func NewSpanAnnotator(req *http.Request, s *trace.Span) *httptrace.ClientTrace {
    spanAnnotator := ochttp.SpanAnnotator{s}

    // register wanted hooks
    return &httptrace.ClientTrace{
        GetConn:              spanAnnotator.PutIdleConn,
        GotFirstResponseByte: spanAnnotator.GotFirstResponseByte,
        DNSStart:             spanAnnotator.DNSStart,
        ...
    }
}

...

transport := ochttp.Transport{
    ClientTracer: NewSpanAnnotator,
}
```

Which reminds me why I wanted SpanAnnotator to embed *trace.Span or otherwise have it be an exported property. Bear in mind that the actual resulting *httptrace.ClientTrace or SpanAnnotator is created and contained within ochttp's traceTransport.RoundTrip().

Reply:
I still don't see why the type needs to be public. You could do this :
```go
t := &ochttp.Transport {
    ClientTracer: func(r *http.Request, s *trace.Span) *httptrace.ClientTrace {
          return &httptrace.ClientTrace {
               GetConn: ann.GetConn
          }
    }
}
```

Reply:
Adding a type adds complexity and distraction to the godoc

Reply:
Yeah I seemed to have forgotten because the function outputs *httptrace.ClientTrace I can always wrap it with a new one. Good thing vacation is coming up for some fresh eyes.

Will not only unexport SpanAnnotator but all its methods can be unexported too.